builtins: fix str to geometry conversion case

otan opened the pull request:
Fixed a bug where if an error with regards to parsing a string to EWKT
was found in a string overload, the error would be overriden by the next
argument resolution.

Refs: #51718

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52245)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52245#-:-MDq8U6O-vnCW5-iT8E0:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 3895 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52245#-MDq8Dfm0L2lGracZv8U:-MDq8Dfn-iTm7goWMCO8:b-reihrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0f344f0f9ad7779791f27dbdd4d0da3406c4e6/pkg/sql/logictest/testdata/logic_test/geospatial#L3895)):*
> ```
> false  false  97  false  false
> 
> # Test for 2+ string arguments for Geometry which one is an error.
> ```

[nit] there's something weird about this sentence. Maybe which -> where?
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3994 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52245#-MDq7I3e87QLXvBu6p2n:-MDq7I3fAw9GaJcpLy4y:blutbbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0f344f0f9ad7779791f27dbdd4d0da3406c4e6/pkg/sql/sem/builtins/geo_builtins.go#L3994)):*
> ```Go
> 		newOverload.Fn = func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
> 			var err error
> 			argsToCast.ForEach(func(i int) {
> ```

Another alternative is to use:
`for i, ok := argsToCast.Next(0); ok; i, ok = argsToCast.Next(i+1) {`

That would allow you to just break early in case of error, and might make it clearer what's going on.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: fix str to geometry conversion case

Fixed a bug where if an error with regards to parsing a string to EWKT
was found in a string overload, the error would be overriden by the next
argument resolution.

Refs: #51718

Release note: None

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52245)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 3895 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52245#-MDq8Dfm0L2lGracZv8U:-MDtxdp12v04pCimVOqy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0f344f0f9ad7779791f27dbdd4d0da3406c4e6/pkg/sql/logictest/testdata/logic_test/geospatial#L3895)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] there's something weird about this sentence. Maybe which -> where?
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 3994 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52245#-MDq7I3e87QLXvBu6p2n:-MDtxe6h6w2XwbaQqkpa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f0f344f0f9ad7779791f27dbdd4d0da3406c4e6/pkg/sql/sem/builtins/geo_builtins.go#L3994)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Another alternative is to use:
`for i, ok := argsToCast.Next(0); ok; i, ok = argsToCast.Next(i+1) {`

That would allow you to just break early in case of error, and might make it clearer what's going on.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented:
tftr!

bors r=rytaft
